File: suchwaiting/src/main/java/com/azalea/suchwaiting/wrapper_drawable/LoadingDrawable.java
package com.azalea.suchwaiting.wrapper_drawable;

import android.graphics.Canvas;
import android.graphics.ColorFilter;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;

import com.azalea.suchwaiting.loadingimplemention.BaseLoadingRender;

import androidx.annotation.IntRange;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;


/**
 * base loading drawable
 */
public class LoadingDrawable extends Drawable implements Drawable.Callback {

	BaseLoadingRender mLoadingRender = null;

	public LoadingDrawable(BaseLoadingRender loadingRender) {
		mLoadingRender = loadingRender;
		mLoadingRender.setDrawableCallBack(this);
	}

	@Override
	public void draw(@NonNull Canvas canvas) {
		mLoadingRender.draw(canvas);
	}

	@Override
	public void setAlpha(@IntRange(from = 0, to = 255) int alpha) {
		mLoadingRender.setAlpha(alpha);
	}

	@Override
	public void setColorFilter(@Nullable ColorFilter colorFilter) {
		mLoadingRender.setColorFilter(colorFilter);
	}

	@Override
	protected void onBoundsChange(Rect bounds) {
		super.onBoundsChange(bounds);
		mLoadingRender.onBoundsChange(bounds);
	}

	@Override
	public int getOpacity() {
		return mLoadingRender.getOpacity();
	}

	@Override
	public void invalidateDrawable(@NonNull Drawable who) {
		invalidateSelf();
	}

	@Override
	public void scheduleDrawable(@NonNull Drawable who, @NonNull Runnable what, long when) {
		scheduleSelf(what, when);
	}

	@Override
	public int getIntrinsicHeight() {
		return mLoadingRender.getIntrinsicHeight();
	}

	@Override
	public int getIntrinsicWidth() {
		return mLoadingRender.getIntrinsicWidth();
	}

	@Override
	public void unscheduleDrawable(@NonNull Drawable who, @NonNull Runnable what) {
		unscheduleSelf(what);
	}
}

File: suchwaiting/src/main/java/com/azalea/suchwaiting/wrapper_drawable/LoadingDrawable.kt
package com.azalea.suchwaiting.wrapper_drawable

import android.graphics.Canvas
import android.graphics.ColorFilter
import android.graphics.Rect
import android.graphics.drawable.Drawable
import androidx.annotation.IntRange
import com.azalea.suchwaiting.loadingimplemention.BaseLoadingRender

/**
 * base loading drawable
 */
class LoadingDrawable(private val mLoadingRender: BaseLoadingRender) : Drawable(), Drawable.Callback {

    init {
        mLoadingRender.setDrawableCallBack(this, this)
        setVisible(visible = false, restart = false)
    }

    override fun setVisible(visible: Boolean, restart: Boolean): Boolean {
        val changed = super.setVisible(visible, restart)
        if (changed) {
            mLoadingRender.setVisible(visible)
        }
        return changed
    }

    override fun draw(canvas: Canvas) {
        mLoadingRender.draw(canvas)
    }

    override fun setAlpha(@IntRange(from = 0, to = 255) alpha: Int) {
        mLoadingRender.setAlpha(alpha)
    }

    override fun setColorFilter(colorFilter: ColorFilter?) {
        mLoadingRender.setColorFilter(colorFilter)
    }

    override fun onBoundsChange(bounds: Rect) {
        super.onBoundsChange(bounds)
        mLoadingRender.onBoundsChange(bounds)
    }

    override fun getOpacity(): Int {
        return mLoadingRender.opacity
    }

    override fun invalidateDrawable(who: Drawable) {
        invalidateSelf()
    }

    override fun scheduleDrawable(who: Drawable, what: Runnable, `when`: Long) {
        scheduleSelf(what, `when`)
    }

    override fun getIntrinsicHeight(): Int {
        return mLoadingRender.intrinsicHeight
    }

    override fun getIntrinsicWidth(): Int {
        return mLoadingRender.intrinsicWidth
    }

    override fun unscheduleDrawable(who: Drawable, what: Runnable) {
        unscheduleSelf(what)
    }
}